bundles/org.eclipse.equinox.p2.transport.ecf/src/org/eclipse/equinox/internal/p2/transport/ecf/RepositoryTransport.java

```diff
@@ -289,33 +289,26 @@ public static DownloadStatus forStatus(IStatus original, URI toDownload) {
 
 	public static DownloadStatus forException(Throwable t, URI toDownload) {
 		if (isForgiveableException(t)) {
-			String value = System.getProperty(TIMEOUT_RETRY);
-			if (value != null) {
-				try {
-					int retry = Integer.valueOf(value).intValue();
-					if (retry > 0) {
-						Integer retryCount = null;
-						if (socketExceptionRetry == null) {
-							socketExceptionRetry = new HashMap<>();
-							retryCount = Integer.valueOf(1);
-						} else {
-							Integer alreadyRetryCount = socketExceptionRetry.get(toDownload);
-							if (alreadyRetryCount == null)
-								retryCount = Integer.valueOf(1);
-							else if (alreadyRetryCount.intValue() < retry) {
-								retryCount = Integer.valueOf(alreadyRetryCount.intValue() + 1);
-							}
-						}
-						if (retryCount != null && retryCount.intValue() <= retry) {
-							socketExceptionRetry.put(toDownload, retryCount);
-							return new DownloadStatus(IStatus.ERROR, Activator.ID, IArtifactRepository.CODE_RETRY,
-									NLS.bind(Messages.connection_to_0_failed_on_1_retry_attempt_2, new String[] {
-											toDownload.toString(), t.getMessage(), retryCount.toString() }),
-									t);
-						}
+			int retry = Integer.getInteger(TIMEOUT_RETRY, 0);
+			if (retry > 0) {
+				Integer retryCount = null;
```

Comment:
Why is an Integer required here?? All these wrapping/unwrapping of int objects look strange and could most likely be replaced maybe by a map.compute even....

Just in case you like to cleanup even more here :-)

Reply:
I spent a couple of minutes here, and the Integer comes from the map (that doesn't allow int) so we have to deal with nullability and so on. This code could definitely be improved, however my quick hack session tells that it can be a can of worm and that it requires more attention to be properly improved than what I'm currently able to spend.

Reply:
Sure no problem, I found map.merge useful in such cases e.g.

`int current = map.merge(key, 1, (a,b)->a+b);`